### ISSUE TITLE
Custom grains from _grains not working on 2014.7.5

### ISSUE THREAD

**Issue by msciciel** [saltstack/salt#22987]
After upgrade from 2014.7.1 to 2014.7.5 minions are not fetching custom grains from master. I'm using gitfs and custom grains are stored under _grains subdirectory.

**Comment by msciciel** [comment]
```
           Salt: 2014.7.5
         Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
         Jinja2: 2.7.2
       M2Crypto: 0.20.2
 msgpack-python: 0.4.4
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: Not Installed
```

**Comment by jfindlay** [comment]
@msciciel, thanks for the report.

**Comment by jfindlay** [comment]
@msciciel, will you try the suggestions in [this comment](https://github.com/saltstack/salt/issues/23009#issuecomment-95794264)?  Thanks.

**Comment by wt** [comment]
@msciciel Hey. Have you tried running the master with using really file_roots instead of gitfs? That seems to work for me. I am having a similar problem with the git pillar.

**Comment by babilen** [comment]
fwiw, custom grains are still working for us on 2015.7.5, example:

```
#!/bin/python

import os.path

def lacp():
   '''
   Test if LACP is enabled.
   '''
   grains = {}
   grains['lacp'] = os.path.exists('/proc/net/bonding/bond0')
   return grains
```
but more complicated ones too. We use salt on Debian with gitpython.

**Comment by wt** [comment]
Here's an example of my salt-master config (which worked in salt 2014.7.1) with the output of the some commands showing the that custom grain is not running: https://gist.github.com/wt/8cbf4787e76318afa3c4

Here's an example of my salt-master config altered to not use gitfs or git_pillar with the output of the same salt call commands showing the the customer grain runs in that case:
https://gist.github.com/wt/bedee85c60e7d8ba7b0a

The source for the custom grain is identical. If you look closely at the not working case, the logs do not include the line about copying the custom grain to the extmod directory in the minion's cache.

**Comment by wt** [comment]
@babilen Which version of GitPython are you running? Here's mine:
```
# yum info GitPython
Loaded plugins: priorities, update-motd, upgrade-helper
980 packages excluded due to repository priority protections
Installed Packages
Name        : GitPython
Arch        : noarch
Version     : 0.3.2
Release     : 0.6.RC1.el6
...
```

**Comment by babilen** [comment]
Sure:
```
root@salt1:~# salt --versions-report
                  Salt: 2014.7.5
                Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.6
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.1.0
                  RAET: Not Installed
                   ZMQ: 3.2.3
                  Mako: Not Installed
 Debian source package: 2014.7.5+ds-1~bpo70+1
# COLUMNS=80 dpkg -l python-git|tail -n1
ii  python-git     0.3.2~RC1-1  all          Python library to interact with Git repositories
```

**Comment by msciciel** [comment]
Without gitfs it's working ok. With gitfs is broken.

We have few environments and plenty of formulas in git so it's quite hard to workaround with cron based pulling from git.  We rely on git branches matches to environments.

**Comment by babilen** [comment]
@msciciel On which distribution does this happen and how do you install salt? The EPEL packages seem to have issues with gitfs which is why I ask.

**Comment by msciciel** [comment]
Centos6 & EPEL

**Comment by msciciel** [comment]
I have reinstalled from this repo: https://copr-be.cloud.fedoraproject.org/results/saltstack/salt/epel-6-x86_64/salt-2014.7.5-1.el5/ but it still doesn't work. Version 2014.7.1 works well with the same package set in the system.

**Comment by wt** [comment]
I am using Amazon Linux with EPEL6.

**Comment by wt** [comment]
There are a lot of packages with older versions for EPEL6. However, given the nature of the failure (that it works when gitfs is disabled), I wonder if some change was made that isn't compatible with the old version of GitPython that ships in epel6.

**Comment by wt** [comment]
So, here is some interesting data. I configured my master to use gitfs backend. Then I tried to saltutil.sync_grains on a minion, which of course failed. Then I did the following:

```
# salt-call cp.list_master
...
    - ../_grains/instance_data.py
...
```

So, the file is there. However, trying to fetch the file directly looks like the following:

```
# salt-call cp.get_file_str salt://_grains/instnace_data.py
/usr/lib64/python2.6/dist-packages/Crypto/Util/number.py:57: PowmInsecureWarning: Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.
  _warn("Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.", PowmInsecureWarning)
[ERROR   ] Unable to cache file 'salt://_grains/instnace_data.py' from saltenv 'base'.
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
IOError: [Errno 2] No such file or directory: ''
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 123, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/__init__.py", line 422, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 227, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 129, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/cp.py", line 295, in get_file_str
    with salt.utils.fopen(fn_, 'r') as fp_:
  File "/usr/lib/python2.6/site-packages/salt/utils/__init__.py", line 1098, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: ''
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 123, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/__init__.py", line 422, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 227, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 129, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/cp.py", line 295, in get_file_str
    with salt.utils.fopen(fn_, 'r') as fp_:
  File "/usr/lib/python2.6/site-packages/salt/utils/__init__.py", line 1098, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: ''
```

So it looks like the file cannot be fetched by the minion.

**Comment by wt** [comment]
And, FWIW, I just realized that my filename for the grain is mispelled, but that is actually how it is on my system.

**Comment by wt** [comment]
Okay, I just configured my master to use the gitfs backend and deleted my minion cache. I made sure that saltutil on the minion fails and tried the following:

```
# salt-call cp.list_master
...
    - ../_grains/instance_data.py
...
```

Then I did the following:

```
# salt-call cp.get_file_str salt://_grains/instance_data.py
...
```

That worked. So the file can be fetched by the minion.

However, saltutil.sync_grains does not work.

**Comment by wt** [comment]
I have found the following difference with respect to behavior on the minon in salt/modules/cp.py.

Configured with gitfs backend:
```
416     return __context__['cp.fileclient'].cache_dir(
417         path, saltenv, include_empty, include_pat, exclude_pat
418     )
```

returns nothing.

Configured with a root, the call to __context__... returns the following:
```
['/var/cache/salt/minion/files/base/_grains/instance_data.py']
```

**Comment by wt** [comment]
Wait, why is there a "../" at the front of the path on the list_master?

**Comment by wt** [comment]
Okay, I just reconfigured my master to use the file root config, and there is no "../" at the front of the results from list_master.

**Comment by wt** [comment]
If I configure my salt-master to use the gitfs backend and start it in the foreground (running "salt-master" from the command line), the list_master results do not include the "../" and the saltutil.sync_grains (from a minion) works.

**Comment by wt** [comment]
Okay, if start salt-master daemonized (with the "-d" command line arg). It includes the "../" at the beginning of the master files and the sync_grains fails.

**Comment by basepi** [comment]
We have figured out the `../` error and @terminalmage submitted a patch today. It will work its way through the branches in the next day or two. It's a super weird issue and may be a bug in `os.path.relpath`, we're not sure yet.

**Comment by terminalmage** [comment]
This appears to be related to some odd behavior of [os.path.relpath()](https://docs.python.org/2/library/os.path.html#os.path.relpath). A fix has been submitted in #23120.

We run ``os.path.relpath()`` to adjust for the [gitfs_root](http://docs.saltstack.com/en/latest/ref/configuration/master.html#gitfs-root) config parameter. The normal behavior of this function will treat the default gitfs_root of ``''`` (empty string) as the current directory, like so:

```python
http://docs.saltstack.com/en/latest/ref/configuration/master.html#gitfs-root
